File: src/main/java/wraith/fabricaeexnihilo/json/other/MeshPropertiesJson.java
package wraith.fabricaeexnihilo.json.other;

import com.google.gson.*;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.modules.sieves.MeshProperties;

import java.lang.reflect.Type;

public final class MeshPropertiesJson extends BaseJson<MeshProperties> {

    private MeshPropertiesJson() {}

    public static final MeshPropertiesJson INSTANCE = new MeshPropertiesJson();

    @Override
    public MeshProperties deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new MeshProperties(context.deserialize(obj.get("id"), IdentifierJson.INSTANCE.getTypeToken()),
                obj.has("enchantability") ? obj.get("enchantability").getAsInt() : 0,
                obj.get("displayName").getAsString(),
                context.deserialize(obj.get("color"), ColorJson.INSTANCE.getTypeToken()),
                context.deserialize(obj.get("keyIngredient"), IdentifierJson.INSTANCE.getTypeToken()));
    }

    @Override
    public JsonElement serialize(MeshProperties src, Type typeOfSrc, JsonSerializationContext context) {
        var json = new JsonObject();
        json.add("id", context.serialize(src.getIdentifier()));
        json.add("enchantability", context.serialize(src.getEnchantability()));
        json.add("displayName", context.serialize(src.getDisplayName()));
        json.add("color", context.serialize(src.getColor()));
        json.add("keyIngredient", context.serialize(src.getKeyIngredient()));
        return json;
    }

}

File: src/main/kotlin/exnihilofabrico/json/MeshPropertiesJson.kt
package exnihilofabrico.json

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import exnihilofabrico.modules.sieves.MeshProperties
import exnihilofabrico.util.Color
import net.minecraft.util.Identifier
import java.lang.reflect.Type

object MeshPropertiesJson: JsonDeserializer<MeshProperties>, JsonSerializer<MeshProperties> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): MeshProperties {
        val obj = json.asJsonObject
        return MeshProperties(
            context.deserialize(obj.get("id"), object:TypeToken<Identifier>(){}.type),
            if(obj.has("enchantability")) obj.get("enchantability").asInt else 0,
            obj.get("displayName").asString,
            context.deserialize(obj.get("color"), object:TypeToken<Color>(){}.type),
            context.deserialize(obj.get("keyIngredient"), object:TypeToken<Identifier>(){}.type))
    }

    override fun serialize(src: MeshProperties, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val json = JsonObject()
        json.add("id", context.serialize(src.identifier))
        json.add("enchantability", context.serialize(src.enchantability))
        json.add("displayName", context.serialize(src.displayName))
        json.add("color", context.serialize(src.color))
        json.add("keyIngredient", context.serialize(src.keyIngredient))
        return json
    }

}